test/python/transpiler/test_basic_mapper.py

```diff
@@ -0,0 +1,219 @@
+        pass_ = BasicMapper(coupling)
+        after = pass_.run(dag)
+
+        self.assertEqual(before, after)
+
+    def test_trivial_in_same_layer(self):
+        """ No need to have any swap, two CXs distance 1 to each other, in the same layer
+         q0:--(+)--
+               |
+         q1:---.---
+
+         q2:--(+)--
+               |
+         q3:---.---
+
+         Coupling map: [0]--[1]--[2]--[3]
+        """
+        coupling = Coupling({0: [1], 1: [2], 2: [3]})
+
+        qr = QuantumRegister(4, 'q')
+        circuit = QuantumCircuit(qr)
+        circuit.cx(qr[2], qr[3])
+        circuit.cx(qr[0], qr[1])
+
+        dag = circuit_to_dag(circuit)
+        before = deepcopy(dag)
+        pass_ = BasicMapper(coupling)
+        after = pass_.run(dag)
+
+        self.assertEqual(before, after)
+
+    def test_a_single_swap(self):
+        """ Adding a swap
+         q0:-------
+
+         q1:--(+)--
+               |
+         q2:---.---
+
+         Coupling map: [1]--[0]--[2]
+
+         q0:--X---.---
+              |   |
+         q1:--|--(+)--
+              |
+         q2:--X-------
+
+        """
+        coupling = Coupling({0: [1, 2]})
+
+        qr = QuantumRegister(3, 'q')
+        circuit = QuantumCircuit(qr)
+        circuit.cx(qr[1], qr[2])
+        dag = circuit_to_dag(circuit)
+
+        expected = QuantumCircuit(qr)
+        expected.swap(qr[2], qr[0])
+        expected.cx(qr[1], qr[0])
+
+        pass_ = BasicMapper(coupling)
+        after = pass_.run(dag)
+
+        self.assertEqual(circuit_to_dag(expected), after)
+
+    def test_keep_layout(self):
+        """After a swap, the following gates also change the wires.
+         qr0:---.--------
+                |
+         qr1:---|--------
+                |
+         qr2:--(+)--[H]--
+
+         Coupling map: [0]--[1]--[2]
+
+         qr0:------.---------
+                   |
+         qr1:--X--(+)--[H]--
+               |
+         qr2:--X-------------
+        """
+        coupling = Coupling({1: [0, 2]})
+
+        qr = QuantumRegister(3, 'q')
+        circuit = QuantumCircuit(qr)
+        circuit.cx(qr[0], qr[2])
+        circuit.h(qr[2])
+        dag = circuit_to_dag(circuit)
+
+        expected = QuantumCircuit(qr)
+        expected.swap(qr[2], qr[1])
+        expected.cx(qr[0], qr[1])
+        expected.h(qr[1])
+
+        pass_ = BasicMapper(coupling)
+        after = pass_.run(dag)
+
+        self.assertEqual(circuit_to_dag(expected), after)
+
+    def test_far_swap(self):
+        """ A far swap that affects coming CXs.
+         qr0:--(+)---.--
+                |    |
+         qr1:---|----|--
+                |    |
+         qr2:---|----|--
+                |    |
+         qr3:---.---(+)-
+
+         Coupling map: [0]--[1]--[2]--[3]
+
+         qr0:-------(+)---.--
+                     |    |
+         qr1:-----X--.---(+)-
+                  |
+         qr2:--X--|----------
+               |  |
+         qr3:--X--X----------
+
+        """
```

Comment:
There seems to be a bug here.
This circuit is not the equivalent to the original. The second swap should be between `qr2` and `qr1` (that's how `qr3` moves up to `qr1` using a chain of swaps). I checked the branch and it does not produce that. Instead it produces this picture. 

Can you see if you can fix?

Reply:
Correct. I fixed it in 49cb295c. Have a look.